projects/plugins/boost/app/assets/src/js/features/speed-score/lib/hooks.ts

```diff
@@ -93,32 +95,24 @@ export const useDebouncedRefreshScore = (
 	{ moduleStates, criticalCssCreated, criticalCssIsGenerating }: RefreshDependencies,
 	loadScore: RefreshFunction
 ) => {
-	const [ scoreConfigString, setScoreConfigString ] = useState(
-		JSON.stringify( [ moduleStates, criticalCssCreated ] )
-	);
+	const lastScoreConfigString = useRef( JSON.stringify( [ moduleStates, criticalCssCreated ] ) );
 
 	// Debounced function: Refresh the speed score if the config has changed.
-	const debouncedRefreshScore = useMemo(
-		() =>
-			debounce( async ( newConfig, oldConfig ) => {
-				if ( oldConfig !== newConfig ) {
-					setScoreConfigString( newConfig );
-					await loadScore();
-				}
-			}, 2000 ),
-		[ loadScore, setScoreConfigString ]
+	// eslint-disable-next-line react-hooks/exhaustive-deps
+	const debouncedRefreshScore = useCallback(
+		debounce( async newConfig => {
+			if ( lastScoreConfigString.current !== newConfig ) {
+				lastScoreConfigString.current = newConfig;
+				await loadScore();
+			}
+		}, 2000 ),
+		[ loadScore ]
 	);
```

Comment:
Non-blocking because this code already exists. But this seems like a hack that's ported over from Svelte days. Is therea  proper React way of doing this instead of JSON.stringify and comparing values as strings? If we're tracking moduleStates and criticalCssCreated values, shouldn't we be tracking the specific values we care about instead? Or use React native tracking methods and not useRef in combination with a JSON.stringify?

Reply:
The individual values are coming in as objects and arrays. Strings are much more convenient for quick comparison. It also helps with not sending a refresh request if the user immediately goes back to the previous module state.

Reply:
This is not how react tracks dependencies. If there's a need to do this in React, then we've refactored something incorrectly in terms of dependency tracking because we can't use the whole object. I can't put my finger on it right now because I lack context here - I'm not sure about the call-tree and what goes where, but looking at this bit of code - it seems like there's a code smell somewhere that we need to address. If you need help, I'm happy to jump on a debugging call and help you get to the bottom of this on Monday.

Reply:
It's a common react problem; objects can be identical in content, but when they are not the same object reference, the effect gets triggered. Common solutions include stringifying, picking deeper properties than the object itself, using memoization hacks to avoid this.

I'm ok with stringifying for this.